genesis/engine/couplers/sap_coupler.py

```diff
@@ -1740,3 +2393,402 @@ def compute_delassus(self, i_p):
+                        clipped_vertices[:, clipped_n_vertices] = polygon_vertices[:, i]
+                        clipped_n_vertices += 1
+                        if distances[j] > 0.0:
+                            wa = distances[j] / (distances[j] - distances[i])
+                            wb = 1.0 - wa
+                            clipped_vertices[:, clipped_n_vertices] = (
+                                wa * polygon_vertices[:, i] + wb * polygon_vertices[:, j]
+                            )
+                            clipped_n_vertices += 1
+                    elif distances[j] <= 0.0:
+                        wa = distances[j] / (distances[j] - distances[i])
+                        wb = 1.0 - wa
+                        clipped_vertices[:, clipped_n_vertices] = (
+                            wa * polygon_vertices[:, i] + wb * polygon_vertices[:, j]
+                        )
+                        clipped_n_vertices += 1
+                polygon_n_vertices = clipped_n_vertices
+                polygon_vertices = clipped_vertices
+
+                if polygon_n_vertices < 3:
+                    # If the polygon has less than 3 vertices, it is not a valid contact
+                    break
+
+            if polygon_n_vertices < 3:
+                continue
+
+            # compute centroid and area of the polygon
+            total_area = gs.EPS  # avoid division by zero
+            total_area_weighted_centroid = ti.Vector([0.0, 0.0, 0.0])
+            for i in range(2, polygon_n_vertices):
+                e1 = polygon_vertices[:, i - 1] - polygon_vertices[:, 0]
+                e2 = polygon_vertices[:, i] - polygon_vertices[:, 0]
+                area = 0.5 * e1.cross(e2).norm()
+                total_area += area
+                total_area_weighted_centroid += (
+                    area * (polygon_vertices[:, 0] + polygon_vertices[:, i - 1] + polygon_vertices[:, i]) / 3.0
+                )
+
+            if total_area < gs.EPS:
+                continue
+            centroid = total_area_weighted_centroid / total_area
+            barycentric0 = tet_barycentric(centroid, tet_vertices)
+            barycentric1 = tri_barycentric(centroid, tri_vertices, normal=candidates[i_c].normal)
+            tangent0 = polygon_vertices[:, 0] - centroid
+            tangent0 /= tangent0.norm()
+            tangent1 = candidates[i_c].normal.cross(tangent0)
+            deformable_g = self.coupler._hydroelastic_stiffness
+            rigid_g = self.coupler.fem_pressure_gradient[i_b, i_e].dot(candidates[i_c].normal)
+            pressure = (
+                barycentric0[0] * tet_pressures[0]
+                + barycentric0[1] * tet_pressures[1]
+                + barycentric0[2] * tet_pressures[2]
+                + barycentric0[3] * tet_pressures[3]
+            )
+            if total_area < gs.EPS or rigid_g < gs.EPS:
+                continue
+            g = 1.0 / (1.0 / deformable_g + 1.0 / rigid_g)  # harmonic average
```

Comment:
avoid division for numerical accuracy and performance. `g = rigid_g * deformable_g / (deformable_g + rigid_g)`